staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go

```diff
@@ -0,0 +1,269 @@
+	return &timingRatioHistogramInner{
+			under:       underMember,
+			numerator:   v.initialNumerator,
+			denominator: initialDenominator,
+		},
+		nil
+}
+
+// WithLabelValuesSafe is the same as WithLabelValuesChecked in cases where that does not
+// return an error.  When the unsafe version returns an error due to the vector not being
+// registered yet, the safe version returns an object that implements its methods
+// by first calling WithLabelValuesChecked and then delegating to whatever observer that returns.
+// In the other error cases the object returned here is a noop.
+func (v *TimingRatioHistogramVec) WithLabelValuesSafe(initialDenominator float64, labelValues ...string) RatioedObserver {
+	tro, err := v.WithLabelValuesChecked(initialDenominator, labelValues...)
+	if err == nil {
+		return tro
+	}
+	if !compbasemetrics.ErrIsNotRegistered(err) {
+		klog.ErrorS(err, "Failed to extract TimingRatioHistogramVec member, using noop instead", "vectorname", v.FQName(), "labelValues", labelValues)
+		return tro
+	}
+	return &timingRatioHistogramInner{
+		under:       &timingHistogramVecElt{under: v.under, labelValues: labelValues},
+		numerator:   v.initialNumerator,
+		denominator: initialDenominator,
+	}
+}
+
+type timingHistogramVecElt struct {
+	under       compbasemetrics.GaugeVecMetric
+	labelValues []string
+}
+
+func (tve *timingHistogramVecElt) Set(x float64) {
+	tve.under.WithLabelValues(tve.labelValues...).Set(x)
+}
+
+// With, if called after this vector has been
+// registered in at least one registry and this vector is not
+// hidden, will return a RatioedObserver that is NOT a noop along
+// with nil error.  If called on a hidden vector then it will
+// return a noop and a nil error.  Otherwise it returns a noop
+// and an error that passes compbasemetrics.ErrIsNotRegistered.
+func (v *TimingRatioHistogramVec) With(initialDenominator float64, labels map[string]string) (RatioedObserver, error) {
+	underMember, err := v.under.WithChecked(labels)
+	if err != nil {
+		return noopRatioed{}, err
+	}
+	underMember.Set(v.initialNumerator / initialDenominator)
+	return &timingRatioHistogramInner{under: underMember,
+			numerator:   v.initialNumerator,
+			denominator: initialDenominator},
+		nil
+}
+
+type noopRatioed struct{}
+
+func (noopRatioed) Observe(float64)        {}
+func (noopRatioed) SetDenominator(float64) {}
+
+func (v *TimingRatioHistogramVec) Reset() {
+	v.under.Reset()
+}
+
+// WithContext returns wrapped TimingHistogramVec with context
+func (v *TimingRatioHistogramVec) WithContext(ctx context.Context) *TimingRatioHistogramVecWithContext {
+	return &TimingRatioHistogramVecWithContext{
+		ctx:                     ctx,
+		TimingRatioHistogramVec: v,
+	}
+}
+
+// TimingHistogramVecWithContext is the wrapper of TimingHistogramVec with context.
+type TimingRatioHistogramVecWithContext struct {
+	*TimingRatioHistogramVec
+	ctx context.Context
+}
+
+type TimingRatioHistogramPairVec struct {
```

Comment:
I found the idea of having one vec returning a pair metrics upon a call to WithLabelValues pretty confusing. From how it was designed in the Prometheus client, WithLabelValues should be used to select one metric of a vec and not multiple based on an assumption of label values. Do you think we could replace that by adding a "request_kind" label in the metric and having an array of metrics.RatioedObserverVec in the controller config? I think it would be closer to what the prometheus client is doing.

Reply:
The distinction between the two members of a pair is not request kind but rather phase of handling --- waiting in a queue vs actively being executed.
The bundling of two gauges into a pair just a convenient bundle because the client uses these in such pairs, it is a natural pairing for this data.  The pairing is built on top of multi-dimensional vectors that break down by handling phase plus other stuff.

Reply:
The ...PairVec interface is really just a wrapper for convenience of the clients in APF, and a small thing at that.  I can think of a couple of alternatives:
1. Augment the `RatioedGaugeInputVec` interface with a method to New.. a pair.
2. Create a function that makes two calls to `RatioedGaugeInputVec::NewForLabelValuesSafe` to create a pair.

Reply:
I think the thing that was confusing me is that the name of the type is very generic but the use case is very local to APF since it is assuming the existence of 2 execution phases. That said while looking at it again, I don't think other alternative implementation would give the same convenience that your wrapper is adding, so I am with it